### ISSUE TITLE
Change Request: Evaluate Codemod to replace config migrator and eslint-transforms

### ISSUE THREAD

**Issue by nzakas** [eslint/eslint#20442]
### ESLint version

HEAD

### What problem do you want to solve?

We currently maintain two different codemod solutions:

- `@eslint/config-migrator` - for migration eslintrc config files to flag config
- `eslint-transforms` - other miscellaneous transforms such as updating rules

We typically don't spend a lot of time on these as they are labor-intensive to change and update. 

### What do you think is the correct solution?

OpenJS Foundation has partnered with [Codemod](https://codemod.com/) to help develop and maintain codemods for its projects.

@alexbit's team has already created some codemods for evaluation and donated them to the project:
https://github.com/eslint/codemods/tree/main/codemods/v9

These are nicely shared in a public registry:
https://app.codemod.com/registry?q=scope%3Aeslint

What's even nicer is that these can be applied to GitHub repos directly, opening PRs with the changes, meaning that people don't have to install a tool to apply the fixes locally and then push to their repo. Examples:

- https://github.com/amirabbas-gh/vercel/pull/8/files
- https://github.com/amirabbas-gh/minigame-demo/pull/5

I'd like to evaluate these against our own solutions and determine if we can switch our docs to use these instead.

### Participation

- [ ] I am willing to submit a pull request for this change.

### Additional comments

Looking for a volunteer to evaluate these codemods against eslint-transforms and config-migrator to see if they are ready to replace them.

**Comment by alexbit-codemod** [comment]
thanks for the write-up. happy to address any questions or concerns, and looking forward to helping ESLint users adopt the latest features with ease.

**Comment by DMartens** [comment]
I would love to evaluate them. As a starting point I would compare common scenarios and then delve into and compare the code between them.

**Comment by alexbit-codemod** [comment]
hey team, any updates here?

**Comment by DMartens** [comment]
I probably made the task bigger than it was intended but I also wanted to check for additional edge cases (especially for the configuration migration).
All in all each codemod has additional features not present in the current transforms but unfortunately there are some bugs.
Below are my notes for whether the codemods are ready to replace our current transforms.
The problems listed below are what I think are what needs to be addressed but other team members may want additional changes given the other differences.
Please feel free to correct me if I made a mistake anywhere.

## New Rule Format
|Feature|codemod|eslint-transforms|
|-------|-------|-----------------|
|Legacy format| (arrow) function expression or any call | (arrow) function expression or known RuleCreator call|
|Module format| ESM and CJS | only CJS|
|Warning for missing options schema | yes | no|
|Advanced rule matching| yes | no |

**Both support**:
- use `schema` from `module.exports.schema` if it exists
- adding `fixable` if `fix` is provided in `context.report`
- custom `context` parameter name

**Result**: Ready to replace `eslint-transforms`.

## V9 Rule Migration

|Feature| codemod | eslint-transforms|
|-|-|-|
|Transformation|-|Each as variable declaration initialized with preferred ?? fallback and replace references|
|Method to Property|untransformed|transformed|
|Clean up|none|removes `const PROPERTY = PROPERTY`|
|getComments|invalid code|warning|
|getScope|TODO comment|inside method|
|Legacy `context.report`|transformed|untransformed|
|`CodePath#currentSegments`|transformed using code from [migration guide](https://eslint.org/blog/2023/09/preparing-custom-rules-eslint-v9/#codepath%23currentsegments) |warning|

**Problems**:
- Properties which were methods (e.g. `getCwd`) are untransformed (this is not mentioned in the migration guide but [here](https://eslint.org/blog/2023/09/preparing-custom-rules-eslint-v9/#context-methods-becoming-properties))
- `getComments` is replaced with string instead of array concatenation of multiple `getComments(Before/Inside/After)` methods

**Result**: Not ready to replace `eslint-transforms`.

## Flat Config
### General
|Feature|codemod|@eslint/migrate-config|
|-------|-------|----------------------|
|Filename|always `eslint.config.mjs`|.eslintrc.* -> eslint.config.*; remaining are kept|
|.gitignore|Included|opt-in via --gitignore|
|File Extensions from package.json#scripts|yes|no|
|Package-specific migration|missing|yes (ava, react-hooks, import, flowtype)|
|Package Instalation|Warning|Warning|
|Output|Always ESM| ESM (default) or CJS; with specific file extension (`.mjs` or `.cjs`)|
|Removes empty override|no|yes|
|Removes eslint/exported comments|yes|no|

### Properties
|Property|codemod|@eslint/migrate-config|
|----|---|---|
|`parser`|also other parsers (e.g. vue or angular)|only `@typescript-eslint`, `babel` and `espree`|
|`root`|removed|removed|
|`env`|appended|prepended|
|`parserOptions`| only object |isObject ? moves `ecmaVersion` + `sourceType` : as-is|
|`plugins`|ignores non-strings in the array|only array of strings, otherwise error; ignored if non-array|
|`ignorePatterns`|Dropped|Warning for non-array; uses `globalIgnores`|
|`globals`|only object|isObject ? appended : as-is|
|`overrides`|skips non-objects|ignores non-array, keeps non-objects as is|
|`files/excludedFiles`|no warning|warning for non-array/non-string|
|`rules`|removes removed rules and migrates options|as-is|
|`processor`, `settings`|as-is| as-is|

**Both support**:
- including `.eslintignore`
- transforming configuration in `.js`, `.json` and `.yaml` format
- define `__dirname` if needed
- migrate `extends` by using `defineConfig`

**Problems**:
- Multiple (e.g. `extends`) properties are dropped when transforming JSON files and only setting that property in the (sub)configuration object
- missing package-specific migrations using [`@eslint/compat`](https://eslint.org/blog/2024/05/eslint-compatibility-utilities/)

**Result**: Not read to replace `@eslint/migrate-config`

**Comment by alexbit-codemod** [comment]
wow, this is a super thorough and helpful analysis. Thank you so much @DMartens. 
**will share it with the team and follow up after we address the issues you called out.**

(qq out of curiosity: do you have any sample PRs from codemod, `eslint-transforms`, or `@eslint/migrate-config` on a real repo? If so, I'd love to look at them to see how the scripts work end to end. Also, do you have any rough sense of false positives and false negatives for the different approaches? For example, out of 100 changes, how many were incorrect and how many were missed. no worries if you don't have this handy. but in general, these kinds of metrics would help us better judge how effective the automation is.)

**Comment by DMartens** [comment]
I do not know about PRs using any of the transformations but you could check the transformations against the before code of popular ESLint plugins (e.g. for [eslint-plugin-vue](https://github.com/vuejs/eslint-plugin-vue/pull/2331/changes)) and check for differences / whether the tests work with ESLint v9.

**Comment by nzakas** [comment]
@alexbit-codemod The test cases in the eslint/rewrite repo under packages/config-migrator are your best source for before-and-after migration examples. Check the fixtures directory.

**Comment by amirabbas-gh** [comment]
Hello @DMartens ,
I'm Amirabbas from Codemod.

First of all, thank you very, very much for your incredibly detailed and thorough report. I carefully read through everything and tried to address the issues you mentioned regarding **V9 Rule Migration** and **Flat Config**, which were a total of four problems. I’ve fixed them in the following commits:

- https://github.com/eslint/codemods/commit/888b771e2c75ba63908d95c1d47703e0b833dc26  
- https://github.com/eslint/codemods/commit/3ed76a7bef99a7b4d321205bc022410ae4e1486a  
- https://github.com/eslint/codemods/commit/d75fc70e21b11b9bcac52201b923f349f7224f27  

Here’s what has been addressed:

- The issue related to transforming certain `context` methods into properties has been fixed.  
- `getComments` is no longer converted into a string; it now correctly becomes an array.  
- `@eslint/compat` is now used for `plugins` and `extends`.

I’ve updated the related test cases accordingly. I’d really appreciate it if you could review these changes when you have time. Thanks again for your effort and detailed feedback.

There’s just one point I’d like to clarify regarding this issue you mentioned:

> Multiple (e.g. extends) properties are dropped when transforming JSON files and only setting that property in the (sub)configuration object

I wasn’t able to create a concrete example to reproduce this behavior, and honestly, I couldn’t observe this issue anywhere in my testing. If you have a minimal reproduction or example, I’d really appreciate it if you could share it.

I did try to make some improvements based on my assumptions, but since I couldn’t reproduce the issue, I wasn’t able to properly verify whether my changes resolve the exact problem you encountered.

Please try the latest version of the codemod, and if the issue is still present, I’d appreciate more detailed information so I can investigate further.

Thanks again 🙏

**Comment by alexbit-codemod** [comment]
@DMartens would you be able to give this another shot? 

https://app.codemod.com/registry/@eslint/v8-to-v9-config

https://app.codemod.com/registry/@eslint/v8-to-v9-custom-rules

thank you so much.
cc @nzakas

**Comment by DMartens** [comment]
I went through the mentioned three commits and the mentioned bugs have been fixed (except the one where I was unclear).
Below are mostly additional notes about potential ESLint v10 support and one bug which was missing from my earlier report.

## Rule Migration
**Additional v10 changes**:  
In the meantime we released v10 which [removes](https://eslint.org/blog/2026/02/eslint-v10.0.0-released/#removed-deprecated-rule-context-members) the `context.parserPath` and `context.parserOptions` properties. It would be great if these would also be included by printing a warning for `parserPath` and updating the path from `context.parserOptions` to `context.languageOptions.parserOptions`.

Further I'd appreciate it if the [removed SourceCode methods](https://eslint.org/blog/2026/02/eslint-v10.0.0-released/#removed-deprecated-sourcecode-methods) will also be transformed (for `getTokenOrCommentBefore`, `getTokenOrCommentAfter`, `isSpaceBetweenTokens`) as mentioned in the linked migration guide.
Further as `getJSDocComment` has no replacement, a warning would be appropriate.

**currentSegments**:  
I think while fixing the context methods to properties bug, a new bug was introduced:
When fixing the code path segments, `context.sourceCode ??` should not be prefixed ([line](https://github.com/eslint/codemods/blob/3ed76a7bef99a7b4d321205bc022410ae4e1486a/codemods/v9/custom-rules/tests/fixtures/expected.js#L41)) before the introduced variable.

## Config Migration
**Missing property**:  
I have to apologize for not being clearer with what I meant with:
```
Multiple (e.g. extends) properties are dropped when transforming JSON files and only setting that property in the (sub)configuration object
```
This problem unfortunately still persists, for example the following `.eslintrc.json`:
```json
{
	"extends": ["eslint-config-airbnb"]
}
```
is transformed into `eslint.config.js`:
```js
import { defineConfig } from "@eslint/config-helpers";

export default defineConfig([]);
```

So the `extends` property is dropped without replacement.
But the transformation works correctly for JavaScript configuration files (`.eslintrc.js`) or by setting an additional property (e.g. an empty `rules` property).

**--target unsupported**:  
Also I forgot to mention another bug: The `--target` CLI option does not work the configuration migrator as the paths in the shell script (`move.sh`) are hardcoded.
I think this would be helpful for bigger projects as some use multiple eslint configurations with the `--config` CLI option.

**Comment by alexbit-codemod** [comment]
thanks so much for detailed review. you rock @DMartens. will make the adjustments and will share updates soon.

**Comment by amirabbas-gh** [comment]
Hi @DMartens,

First of all, thank you again for your careful and detailed review, I truly appreciate it.

Regarding the changes introduced in ESLint v10 that you mentioned: the codemod we are currently working on (and trying to bring to a stable release) is specifically focused on migrating from **v8 to v9**. I believe it’s more structured and principled to first complete this codemod strictly for that responsibility and introduce it in the ESLint upgrade guide for migrating from version 8 to version 9.
Once we are fully confident about the stability of these codemods, we can then implement a separate codemod dedicated to migrating from **v9 to v10**, develop it independently, and ensure its stability as well. After that, we could publish a workflow that runs both codemods sequentially (v8 → v9 and v9 → v10), enabling a safe and reliable migration path from **v8 to v10**.

Regarding the issues you raised:
- The `extends` issue in JSON has been resolved.
- The `currentSegment` issue in custom rules has been fixed, and I also updated its test case to make the behavior clearer for users.
- The `--target` issue has been resolved as well — and thank you for pointing that out.

Thanks again for your thoughtful feedback.

[Custom rule migration codemod publish](https://github.com/eslint/codemods/actions/runs/22465637744)
[Config file migration codemod publish](https://github.com/eslint/codemods/actions/runs/22465637744)

https://app.codemod.com/registry/@eslint/v8-to-v9-config
https://app.codemod.com/registry/@eslint/v8-to-v9-custom-rules

**Comment by DMartens** [comment]
Thank you for the quick turnover. Of course the ESLint v10 support can wait, so we can focus on correctness first.
I checked the three fixes but unfortunately two of them are only partly fixed:
- The issue with the single property was not specific to `extends` but also other properties (e.g. `languageOptions` or `settings`).
- `--target` does the transform but there are two problems:
  - Because of the code formatting with prettier fails ([error] No files matching the pattern were found: "**/eslint.config.mjs" is printed in the console)
  - The transformation may stringify the input (see example below)

```javascript
// input: file.js
export default {
	rules: {
		'semi': 'error',
	}
};

// output:
import { defineConfig, globalIgnores } from "@eslint/config-helpers";

export default defineConfig([
  globalIgnores([
    "export default {",
    "	rules: {",
    "		'semi': 'error',",
    "	}",
    "}"
  ]),
]);
```

**Comment by alexbit-codemod** [comment]
thanks @DMartens. heads up, given @amirabbas-gh's availability, it might take some time for him to get to fixing these final issues.

some idea: if you think this codemod, in its current state, is still beneficial to users, how about we mention it in the docs but also call out the known issues? this way we can surface more gaps as more users try it, and address those along with the current issues in upcoming versions. whatever decision you and @nzakas make here works for us. just wanted to share this idea as an option. thanks again for your time reviewing the codemod.

**Comment by nzakas** [comment]
Sorry, I've not had time to follow this thread. To me the primary concern is ensuring that this codemod correctly mirrors the functionality we already have in `@eslint/config-migrator`. If it does, then I'm happy to add it to the docs; if not, then we should wait until it does. I'll leave it up to @DMartens to make that assessment.

**Comment by alexbit-codemod** [comment]
great point. thank you and huge thanks to @DMartens for the help with this task.

**Comment by DMartens** [comment]
I do not think the codemod for the config transformation is quite there yet (see the two bugs above). I agree that we should wait as there is no rush to finish this.

**Comment by alexbit-codemod** [comment]
noted. will ask the folks in our community to see if someone can wrap this quickly while amirabbas is out. thank you.

**Comment by amirabbas-gh** [comment]
Hi @DMartens,

Thanks for your patience, and thanks again for your review and for pointing out the gaps so we can make these codemods more useful for the community. I was also away from open-source contributions for a bit due to some personal situations.

Here’s a detailed update on what’s been addressed and improved:

1. **Single-property detection issue**  
   I investigated the bug and found that single properties in both YAML and JSON were not being detected properly, so they weren’t going through the migration process. This is now fixed—regardless of whether `.eslintrc.json` or `.eslintrc.yaml` contains one or multiple properties, everything is correctly picked up and migrated.

2. **Prettier step improvements**  
   Fixed the issue where the workflow would fail if no files were found. It now exits gracefully without throwing errors.  
   I also introduced a new workflow parameter called `codeFormattingCommandEnabled`, clients can use them using `--param`:
   - `codeFormattingCommandEnabled` (boolean, default: `false`): Enables/disables the formatting step.
   - `codeFormattingCommand` (string, default: `npx prettier --write "**/eslint.config.mjs" --ignore-path /dev/null --no-config --no-error-on-unmatched-pattern`): Command to run when formatting is enabled.  
   These are now documented in the README.

3. **Stringify issue**  
   This turned out to only affect JavaScript-based `.eslintrc` files using `export default`. That issue has now been resolved.

4. **Test improvements**  
   Updated tests for both `config-file` and `custom-rules` codemods so they can be executed via `npm test`. This makes it easier to validate correctness and ensure the codemods run reliably.

5. **Using the `rename` feature**  
   Refactored file renaming and `.eslintignore` removal to use the Codemod `rename` feature:  
   https://docs.codemod.com/jssg/advanced#root-rename-%E2%80%94-rename-the-current-file  
   This improves both safety and code cleanliness.


[Custom rule migration codemod publish](https://github.com/eslint/codemods/actions/runs/22465637744)
[Config file migration codemod publish](https://github.com/eslint/codemods/actions/runs/22465637744)

https://app.codemod.com/registry/@eslint/v8-to-v9-config
https://app.codemod.com/registry/@eslint/v8-to-v9-custom-rules

**Comment by DMartens** [comment]
Thank you for continuing to work on this and I hope you are doing well.
I am happy to confirm that the 3 bugs (stringify, single property and prettier) have been fixed.
Unfortunately I found some additional bugs / possible improvements while confirming the fixes:
- The [`linterOptions`](https://eslint.org/docs/v8.x/use/configure/migration-guide#linter-options ) `noInlineConfig` and `reportUnusedDidsableDirectives` are currently unsupported
- When converting a configuration with `extends` it adds the message that "path" and "url" have to be installed but I think the execution environments (Node.js, Bun, Deno) have them already
- The `--target` CLI option is no longer working as the rename assumes it is the target directory. Maybe I misused it to pass the configuration file name. Either way passing the configuration file name should be supported as ESLint allows passing a custom configuration in with the `--config` CLI option

If you want to go the extra mile, ESLint had support for passing the configuration in [`eslintConfig` inside `package.json`](https://eslint.org/docs/v8.x/use/configure/configuration-files#:~:text=package.json%20%2D%20create%20an%20eslintConfig%20property%20in%20your%20package.json%20file%20and%20define%20your%20configuration%20there.)

**Comment by amirabbas-gh** [comment]
Hi @DMartens,

I’m really happy that I was able to fix those three bugs, and thanks again for your patience and for reviewing everything so carefully. I think we’re getting very close to a stable solution that can handle everything reliably. Let me walk you through the recent updates:

- `linterOptions` are now properly migrated and this is fully handled.

- Regarding `path` and `url`: I understand your concern. Since the config file is transformed and new imports may be introduced, I added a log to suggest these dependencies to ensure nothing breaks for the user running the codemod.  
  Removing `path` and `url` from that log is easy, but I’d argue it’s not a big deal to keep them there—it gives us extra safety and confidence that the migration runs 100% reliably.

- `--target` is only meant to specify the target directory. For your use case, I’ve introduced two new params:
  - `eslintConfigCustomName`: lets you define the config file name
  - `eslintConfigLanguage`: specifies the config format (`javascript`, `yaml`, or `json`) — defaults to `javascript`

  Example:
  ```bash
  npx codemod@latest @eslint/v8-to-v9-config \
    --param eslintConfigCustomName=.amirabbas.eslintrc.test.json \
    --param eslintConfigLanguage=json


- Support for migrating eslintConfig inside package.json has also been added. You can check the [test cases here](https://github.com/eslint/codemods/tree/main/codemods/v9/config-file/tests).

**Comment by DMartens** [comment]
I agree that this is shaping up to be a substantial improvement over the existing migrators and thank you for going the extra mile with the `package.json` support.
Happy to confirm that all existing bugs have been squashed.
I think the next step is to to compare the output of the codemods with the existing tests of the config and rule migrators.
I will report back tomorrow with my findings.

**Comment by amirabbas-gh** [comment]
Thanks @DMartens, let’s go for it 🎉 Looking forward to your report!

**Comment by DMartens** [comment]
Here are my results of comparing the codemods with the fixture tests of the migrators.
Good news first, most of the output is equal (✔️) and failing fixtures usually expose a single small bug.
Further each codemod has one shared bug which occurred in multiple tests (🐛).

## Config 
The shared bug for the config migrator is that invalid plugin identifiers are not quoted, e.g.:
```javascript
export default {
	plugins: {
		@typescript-eslint: tsPlugin,
	}
}
```

|Fixture|Status|Notes|
|-------|------|-----|
|[`backend-nx-skeleton-esm-js`](https://github.com/eslint/rewrite/tree/main/packages/migrate-config/tests/fixtures/backend-nx-skeleton-esm-js)| ✔️ |Relative paths in `extends` also use the `FlatCompat`|
|[`backend-nx-skeleton-js`](https://github.com/eslint/rewrite/tree/main/packages/migrate-config/tests/fixtures/backend-nx-skeleton-js)| ✔️ |converted to `eslint.config.mjs` from CommonJS|
|[`basic-eslintrc`](https://github.com/eslint/rewrite/tree/main/packages/migrate-config/tests/fixtures/basic-eslintrc)| 🐛| `@typescript-eslint` is not quoted inside `plugins`, plugins are imported multiple times|
|[`formidable-js`](https://github.com/eslint/rewrite/tree/main/packages/migrate-config/tests/fixtures/formidable-js)| ❌| codemod drops local variables used in the config|
|[`gitignore-complex/simple`](https://github.com/eslint/rewrite/tree/main/packages/migrate-config/tests/fixtures/gitignore-complex)| ✔️|codemod inlines the patterns while config migrator reads the file|
|[`import-duplicate-js`](https://github.com/eslint/rewrite/tree/main/packages/migrate-config/tests/fixtures/import-duplicate-js)| 🐛|plugin key "react-hooks"|
|[`no-globals-for-env`](https://github.com/eslint/rewrite/tree/main/packages/migrate-config/tests/fixtures/no-globals-for-env)|🐛|
|[`override-extends`](https://github.com/eslint/rewrite/tree/main/packages/migrate-config/tests/fixtures/overrides-extends)| ✔️ |`files` use a string (not a string array)|
|[`plugins-dedupe`](https://github.com/eslint/rewrite/blob/main/packages/migrate-config/tests/fixtures/plugins-dedupe)| ✔️|
|[`prisma-js`](https://github.com/eslint/rewrite/tree/main/packages/migrate-config/tests/fixtures/prisma-js)| ❌| missing local ignores (only from `.eslintignore` but not from config (`ignorePatterns: ["tmp"]`))|
|[`release-it`](https://github.com/eslint/rewrite/blob/main/packages/migrate-config/tests/fixtures/release-it)| ❌| `eslint:recommended` missing (must also be passed to `compatWithRecommended.extends`)|
|[`reveal-md`](https://github.com/eslint/rewrite/tree/main/packages/migrate-config/tests/fixtures/reveal-md)| ✔️|`.eslintrc` not supported (but works with explicit `.json` file extension)|
|[`slash-package`](https://github.com/eslint/rewrite/tree/main/packages/migrate-config/tests/fixtures/slash-package)|🐛| |

## Rule
For the rule migrator the shared bug is that object methods for `create are not detected, e.g.:
```javascript
export default {
	create(context) { // only (arrow) function expression are detected
		context.getSourceCode();
	}
}
```

### Rule Format
|Fixture|Status|Notes|
|-------|------|-----|
|[already transformed](https://github.com/eslint/eslint-transforms/blob/main/tests/fixtures/lib/new-rule-format/already-transformed.input.js)| ✔️ | |
|[arrow-function](https://github.com/eslint/eslint-transforms/blob/main/tests/fixtures/lib/new-rule-format/arrow-function.input.js)| ✔️ ||
|[custom-identifier-context](https://github.com/eslint/eslint-transforms/blob/main/tests/fixtures/lib/new-rule-format/custom-identifier-context.input.js)| ✔️ | |
|[fixable](https://github.com/eslint/eslint-transforms/blob/main/tests/fixtures/lib/new-rule-format/fixable.input.js)| 🐛 |  |
|[no-schema](https://github.com/eslint/eslint-transforms/blob/main/tests/fixtures/lib/new-rule-format/no-schema.input.js)| ✔️| migrator does not add `schema: []` (I prefer codemod always adding the `meta.schema`)
|[schema-uses-variable](https://github.com/eslint/eslint-transforms/blob/main/tests/fixtures/lib/new-rule-format/schema-uses-variables.input.js)| ✔️| migrator also moves top-level variables before the rule definition
|[simple](https://github.com/eslint/eslint-transforms/blob/main/tests/fixtures/lib/new-rule-format/simple.input.js)| ✔️ | |
|[with-comments](https://github.com/eslint/eslint-transforms/blob/main/tests/fixtures/lib/new-rule-format/with-comments.input.js)| ✔️ | migrator moves comment above the schema variable |
|[wrapped-in-function](https://github.com/eslint/eslint-transforms/blob/main/tests/fixtures/lib/new-rule-format/wrapped-in-function.input.js)| ? | untransformed |

### V8 -> V9
|Test|Status|Notes|
|-------|------|-----|
|[migrate deprecated context methods to properties](https://github.com/eslint/eslint-transforms/blob/c272756b3ada3ac65bdfcdb4cfdd9b00250d7cb8/tests/lib/v9-rule-migration/v9-rule-migration.test.js#L17)| 🐛| |
|[#2](https://github.com/eslint/eslint-transforms/blob/c272756b3ada3ac65bdfcdb4cfdd9b00250d7cb8/tests/lib/v9-rule-migration/v9-rule-migration.test.js#L70)| 🐛 | |
|[migrate depracated context methods to SourceCode](https://github.com/eslint/eslint-transforms/blob/c272756b3ada3ac65bdfcdb4cfdd9b00250d7cb8/tests/lib/v9-rule-migration/v9-rule-migration.test.js#L108)| ❌| does not support chaining on the created source code (e.g. `context.getSourceCode().getText()`, see variable declarations inside `FunctionDeclaration`)|
|[signature change](https://github.com/eslint/eslint-transforms/blob/c272756b3ada3ac65bdfcdb4cfdd9b00250d7cb8/tests/lib/v9-rule-migration/v9-rule-migration.test.js#L227)| ❌ |does not use the parameter name for `getAncestors`,  does not pass the node to `sourceCode.getDeclaredVariables` |
|[context.getComments](https://github.com/eslint/eslint-transforms/blob/c272756b3ada3ac65bdfcdb4cfdd9b00250d7cb8/tests/lib/v9-rule-migration/v9-rule-migration.test.js#L282)|✔️| unsupported in the migrator |
|[codePath.currentSegments](https://github.com/eslint/eslint-transforms/blob/c272756b3ada3ac65bdfcdb4cfdd9b00250d7cb8/tests/lib/v9-rule-migration/v9-rule-migration.test.js#L299)|✔️| unsupported in the migrator |

**Comment by amirabbas-gh** [comment]
Hi @DMartens, thanks for your review and your report. I'm working on it, and I'll update you here as soon as they are fixed.

**Comment by amirabbas-gh** [comment]
Hi @DMartens,

thanks again for your thorough report — it was genuinely helpful.

I went through all reported issues in detail and focused not only on fixing them, but also on validating potential side effects. I re-ran and cross-checked both **config-file codemod** and **custom rule codemod** against all the fixtures you listed (including the ones already passing ✔️) to ensure consistency and avoid regressions.

Additionally, I added new test cases for previously failing scenarios and shared bugs.

---

## ✅ Summary of Fixes

### Config-file codemod

| Area | Fix |
|------|-----|
| Invalid identifiers | Properties with special characters (e.g. `@`, `/`, etc.) are now correctly quoted across **plugins**, **languageOptions**, and other sections |
| Duplicate imports | Deduplication of imports is now handled properly |
| Local variables | No longer dropped — now resolved via [semantic analysis](https://docs.codemod.com/jssg/semantic-analysis) |
| `files` format | Always normalized to an array of strings |
| `ignorePatterns` | Fully supported and migrated correctly |
| `eslint:recommended` / `eslint:all` | No longer dropped during migration |
| `.eslintrc` support | Now properly supported and migrated |

**Key Improvement:**  
Using Codemod’s semantic analysis  
https://docs.codemod.com/jssg/semantic-analysis  
we now detect variables referenced inside config files and preserve them during migration (fixes issues like `formidable-js`).

---

### Custom-rule codemod

| Area | Fix |
|------|-----|
| `create` method detection | Object method syntax (not just arrow/function expressions) is now supported |
| Wrapped rules | Rules wrapped inside functions are now migrated correctly |
| SourceCode chaining | Supports chained calls like `context.getSourceCode().getText()` |
| API migration accuracy | Fixed `getAncestors` parameter handling and `sourceCode.getDeclaredVariables(node)` usage |

---

## 🐛 Previously Reported Issues — Status

| Category | Status |
|----------|--------|
| Shared config bugs | ✅ Fixed |
| Shared rule bug (`create` detection) | ✅ Fixed |
| V8 → V9 chaining issue | ✅ Fixed |
| `getAncestors` / declared variables issues | ✅ Fixed |
| Wrapped rule transformation | ✅ Supported |

---

## 🧪 Added / Updated Test Cases

I added dedicated test cases for the failing fixtures you highlighted:

- https://github.com/eslint/codemods/tree/main/codemods/v9/config-file/tests/transform-eslintrc-yaml-basic-eslintrc  
- https://github.com/eslint/codemods/tree/main/codemods/v9/config-file/tests/transform-eslintrc-javascript-formidable-js  
- https://github.com/eslint/codemods/tree/main/codemods/v9/config-file/tests/trasnfrom-eslintrc-yaml-no-globals-for-env  
- https://github.com/eslint/codemods/tree/main/codemods/v9/config-file/tests/transform-eslintrc-javascript-prisma-js  
- https://github.com/eslint/codemods/tree/main/codemods/v9/config-file/tests/transform-eslintrc-json-slash-package  

For custom rules, I incorporated nearly all fixtures you shared to ensure full coverage:

- https://github.com/eslint/codemods/tree/main/codemods/v9/custom-rules/tests

---

## 🔍 Final Note

I intentionally revalidated even the previously passing cases to ensure these fixes didn’t introduce regressions. The goal here wasn’t just patching bugs, but making the migration more reliable overall.

If you spot anything else or want me to double-check a specific edge case, I’m happy to dig deeper.

Thanks again for the detailed feedback 🙌

**Comment by DMartens** [comment]
Thank you for the quick turnaround.
Happy to report that I can confirm the fixes.
Except that the fix for `.eslintrc` (without file extension) is incorrect, as it can be any supported file format (not only JSON).
I think this would be better handled by documenting this as a use case of the available parameters of the codemod:
```bash
npx codemod @eslint/v8-to-v9-config eslintConfigCustomName=.eslintrc eslintConfigLanguage=json|yaml|js`
```
I marked that test as passed as the config file without file extension was deprecated a long time ago.  

I think the next step would be to let other people test this; either first from the ESLint team or directly via aforementioned beta test.

**Comment by amirabbas-gh** [comment]
Hi @DMartens,

I see your point and agree. I’ve removed .`eslintrc` (without an extension) as a default pattern from the migration workflow. Folks can still handle this case using the available parameters if needed.

**Comment by alexbit-codemod** [comment]
Thanks @DMartens for the detailed review of the codemods over several iterations. Can we help with the next steps you mentioned, or will you also be kind enough to send it through more iterations by others?

**Comment by DMartens** [comment]
I messaged the team internally to gather additional feedback and would wait 1-2 week(s) for this to finish.
After that I would propose a blog post on the ESLint website for a public beta test.
While checking the documentation for the team message, I found two small problems:
- [v8-to-v9-custom-rules](https://app.codemod.com/registry/@eslint/v8-to-v9-custom-rules) was not recently updated so misses the last batch of changes
- [v8-to-v9-config](https://app.codemod.com/registry/@eslint/v8-to-v9-config) documentation does not mention the added parameters `eslintConfigCustomName` and `eslintConfigLanguage`

**Comment by amirabbas-gh** [comment]
Hi @DMartens, thanks for your attention to the `@eslint/v8-to-v9-custom-rules` codemod.
I had forgotten to release the new version of this codemod, but it’s now published.
Also, the `@eslint/v8-to-v9-config` README.md file has been updated to reflect the recent changes.

**Comment by amirabbas-gh** [comment]
Hi @DMartens, just checking in to see whether there have been any additional updates or feedback from the team on the public beta proposal for the ESLint website.
Happy to help with anything needed from our side, and really appreciate you continuing to drive this forward.

**Comment by DMartens** [comment]
Thanks for checking in.
The team did not find additional bugs. I did want to check edge cases and available rules / configs but unfortunately did not have time for yet. I'll try to do this tomorrow.

**Comment by DMartens** [comment]
Sorry for the delay.
Here are the results when checking with real-world configuration and rule files (sourced from publicly available GitHub repositories) and checking for edge-cases manually.
What I mentioned as "polish" and "suggestion" would be nice-to-have but bugs should be fixed.

## Config
- bug: custom configuration file paths passed via the parameter `eslintConfigCustomName` should keep their name and not be renamed to `eslint.config.mjs` (currently existing `eslint.config.mjs will also be overridden when transformaing a custom config)
- polish: the codemod adds unnecessary `languageOptions` with empty `globals` and `parserOptions` (e.g. by only setting non-empty `rules`)

## Rule
- ESLint v10 [changes](https://eslint.org/blog/2026/02/eslint-v10.0.0-released/#removed-deprecated-rule-context-members) (which were postponed earlier)
  - `context.parserOptions` to `context.languageOptions.parserOptions`
  - print a warning for using the removed `context.parserPath`
  - migrate the removed `SourceCode` methods according to the [migration guide](https://eslint.org/blog/2026/02/eslint-v10.0.0-released/#removed-deprecated-sourcecode-methods) 
  - a warning for `getJSDocComment` as it has no replacement
- bug: does not add a node parameter for a visitor for `context.getScope` (e.g. `Program() { context.getScope() }`)
- bug: not all converted method calls are wrapped in parenthesis when chaining a method (e.g. `context.getFilename().endsWith(".ts")`)
- bug: the detection for wrapping rules in a function call has to be stricter ([example](https://github.com/vuejs/eslint-plugin-vue/blob/master/lib/rules/comma-style.js)) (maybe only if it is the first or only argument?)
- polish: indirectly exported rule functions are not detected (e.g. `function rule(context) {...} module.exports = rule;`)
- suggestion: add a TODO comment when `context` is passed to a function

After fixing the mentioned bugs, I think this is read for a public beta.
Thank you for your continued work on this.

**Comment by jookira** [comment]
I think a transform that migrates `includeIgnoreFile` from `@eslint/compat` to `eslint/config` would be a good addition to the ESLint v10 changes mentioned by @DMartens.

**Comment by alexbit-codemod** [comment]
Thanks @DMartens for the detailed review of the codemods. The current effort is focused on v8 → v9, (not v10). 

Once we fix the v8 → v9 bugs and release it publicly, we and the community can work on v9 → v10. In the future, users can then daisy-chain both recipes to go from 8 → 10. Just want to make sure we stay aligned and focused on v8 → v9 for now. thanks. cc @amirabbas-gh

**Comment by amirabbas-gh** [comment]
Hi @DMartens, thanks again for the re-review. I really appreciate how carefully and thoroughly you go through everything, it genuinely helps us end up with a codemod that has very few issues, or hopefully none at all!

Here’s a recap of the completed changes:

### Config file codemod

* Fixed the bug related to config files with custom names
* Applied the polish suggestion you mentioned and updated the related test cases

### Custom rule codemod

* Fixed the `context.getScope` bug by properly passing the `node`
* Fixed this bug: not all converted method calls were wrapped in parentheses when chaining methods (e.g. `context.getFilename().endsWith(".ts")`)
* Fixed this issue as well: detection for wrapping rules in a function call now has stricter logic (`maybe only if it is the first or only argument?`)
  * Added two dedicated test cases to make sure it’s covered properly
* Added this polish improvement too:
  * indirectly exported rule functions are now detected (e.g. `function rule(context) {...} module.exports = rule;`)

I think with these changes everything should finally be fully ready — let’s go 🚀

**Comment by DMartens** [comment]
Thank you for the quick turnaround.
I checked the fixes and can again confirm all of them fixed.
As I said I think the next step would be a open beta, but I'll leave this for @nzakas to decide.

**Comment by alexbit-codemod** [comment]
@nzakas, wanted to check in and see if there are any updates. Let us know if you have any questions, concerns, or need anything else wrt the eslint codemods. thank you

**Comment by nzakas** [comment]
We discussed in today's TSC meeting. We think the best thing to do is to invite you @alexbit-codemod to write a guest post on the ESLint blog announce the availability. See the details here:

https://github.com/eslint/eslint.org/tree/main/src/content/drafts

I will caution -- I expect the response to be fairly tepid given that this is only v8 -> v9 and I think anyone who wants to upgrade to v9 will have probably already done so. A better announcement would be a v8 -> v10 codemod. 

And please be sure you're pinging the whole TSC when necessary. I'm not around much these days so you'll get a faster response by contacting the whole team.

**Comment by alexbit-codemod** [comment]
noted. thank you @nzakas. sure, we’ll draft the blog post and keep you and the TSC team posted.

i understand the v8 to v9 effort got a bit prolonged and folks are now looking forward to v10. but to keep the scope manageable and avoid further delays, i suggest we wrap this up and move forward with the announcements.

future codemods should hopefully move significantly faster since the foundation will already be in place and more folks in the community can jump in and help with the remaining codemods. and as mentioned earlier, codemods can be easily daisy-chained to take folks from v8 to v10.

as soon as we announce v8 to v9, I encourage folks in the community to contribute v9 to v10 codemods. stay tuned for the announcement blog post.

**Comment by sam3k** [comment]
The TSC agreed on 2026-05-28 to invite the Codemod team to write a guest blog post publicizing the availability of the v8-to-v9 config codemod.

**Comment by lumirlumir** [comment]
A quick question about it:

What would be the future of the `@eslint/migrate-config` package if we decide to use Codemod as our primary code migration solution?

There’s currently an open PR related to this decision, so I’m wondering. https://github.com/eslint/rewrite/pull/457

**Comment by alexbit-codemod** [comment]
is this the same thing as https://github.com/eslint/codemods/tree/main/codemods/v9/config-file

even if not, we can port that over to a jssg codemod and put it in eslint/codemods repo to have a unified place for all things codemod and migration scripts. 

btw, thanks to @khadatkarrohit, v9 to v10 codemods PR are also ready, one already merged:
https://github.com/eslint/codemods/pulls?q=is%3Apr+label%3Av9-to-v10

(cc some eslint codemod champs who recently made contributions @amirabbas-gh , @DMartens , @khadatkarrohit)

**Comment by khadatkarrohit** [comment]
> is this the same thing as https://github.com/eslint/codemods/tree/main/codemods/v9/config-file
> 
> even if not, we can port that over to a jssg codemod and put it in eslint/codemods repo to have a unified place for all things codemod and migration scripts.
> 
These are two separate migrations. The existing v9/config-file codemod handles the v8→v9 flat config migration (.eslintrc → eslint.config.mjs).

[PR #12](https://github.com/eslint/codemods/pull/12) (v9-to-v10-config) targets the v9→v10 migration and removes features that ESLint v10 now errors on, such as eslint-env comments, ESLINT_USE_FLAT_CONFIG, and deprecated CLI flags.

The v9→v10 migration consists of four codemods in total ([#8 is merged, while #9, #12, and #13 are currently under review](https://github.com/eslint/codemods/pulls)).

Once those are merged, we can look at providing a unified v8→v10 migration workflow that runs both migration paths sequentially. thought?

**Comment by alexbit-codemod** [comment]
for the record of this issue, chatted with @lumirlumir. given that our new v8-to-v9 codemods cover  `@eslint/migrate-config` and `eslint-transforms` we can deprecate those and have a unified place moving forward for everything codemod/migration related. 

also, thanks to @khadatkarrohit for v9-to-v10 codemods and @lumirlumir and @DMartens for reviewing the open PRs, soon we will publish those too and include both in the announcement post. stay tuned.
